### PR TITLE
[#6992] Adds administration UI for tags

### LINKED ISSUE #6992: Add admin tag pages for taggable records
It would be useful to have:

* `/admin/bodies/tags`
* `/admin/requests/tags`
* etc

To serve a few purposes:

* Make it clear what tags even exist (https://github.com/mysociety/whatdotheyknow-theme/issues/1158)
* Provide some stats on how many times each tag is used (e.g. https://github.com/mysociety/whatdotheyknow-theme/issues/1159)
* Link to searches for content tagged with a given tag (e.g. click tag `takedown` to see all tagged records of the current taggable with the `takedown` tag)

A basic listing would be a good first step, but additional niceties would be:

* search
* filters (sort by name, most used)
* a `both_links` implementation so that we get `<eye> tag`, where the eye links to a public search for the tag and the tag links to an admin search for the tag.
* A way of finding _all_ taggables with the given tag (I guess a link to search at an overall `/admin/tags` page)

### PR DESCRIPTION
**Pull request by gbp**
## Relevant issue(s)

Fixes #6992

## What does this do?

Adds administration UI for tags

## Why was this needed?

We need somewhere to add generalised notes to a tag #6994 which means we an admin UI for tags.

## Screenshots

/admin/bodies/tags
Screenshot 2022-08-03 at 12-56-55 Authority tags - WhatDoTheyKnow admin

/admin/requests/tags
Screenshot 2022-08-03 at 12-57-03 Request tags - WhatDoTheyKnow admin

/admin/snippets/tags
Screenshot 2022-08-03 at 12-57-09 Snippet tags - WhatDoTheyKnow admin

/admin/bodies/tags?query=coins
Screenshot 2022-08-03 at 12-57-23 Authority tags - WhatDoTheyKnow admin

/admin/snippets/tags?query=:s12
Screenshot 2022-08-03 at 12-57-35 Snippet tags - WhatDoTheyKnow admin

/admin/bodies/tags/department
Screenshot 2022-08-03 at 12-58-00 Authority tags – department - WhatDoTheyKnow admin

/admin/bodies/tags/department?query=dpr:Z
Screenshot 2022-08-03 at 12-58-27 Authority tags – department - WhatDoTheyKnow admin

/admin/bodies
Screenshot 2022-08-03 at 12-58-37 Listing public authorities - WhatDoTheyKnow admin

/admin/requests
Screenshot 2022-08-03 at 12-58-47 Listing FOI requests - WhatDoTheyKnow admin

/admin/snippets
Screenshot 2022-08-03 at 12-58-57 Listing Snippets - WhatDoTheyKnow admin

### CODE REVIEW

**Review by garethrees — CHANGES_REQUESTED** [review]
Nice progress!

As noted in more detail inline, there are two main things that we should take a look at.

1. Taggable is polymorphic, and we want to make heavy use of it across many models in future. The current implementation means we'd need to do quite a bit of work each time we add Taggable to a new class. It's not _essential_ to fix this now, but I'm loath to introduce cruft in a new feature where I don't think we're working around existing cruft.
2. More of an issue is `Admin::TagsController#show`. I've laid out my thoughts inline so no need to repeat here, but I think this is a a more essential fix to make before merge. The main thing to note here is that the existing implementation goes too far for a more incidental part of the project. A really basic version would have done for now – just enough to have somewhere to render a form – and then these flourishes can be added _after_ the core functionality is working.

File: app/helpers/admin/tag_helper.rb
Comment (gbp) [reply app/helpers/admin/tag_helper.rb]: Commit [43c8a99](https://github.com/mysociety/alaveteli/pull/7201/commits/43c8a99738cdddf5eda4c34445c1f7f6fb2122e2) introduces `Taggable.models` which is updated when the concern is included into a model so we don't have to manually update this hash.

File: app/helpers/admin/tag_helper.rb
Comment (gbp) [reply app/helpers/admin/tag_helper.rb]: These are not needed after https://github.com/mysociety/alaveteli/pull/7201/commits/8177010ba046e48d9030a7f5f83a97eb45bda584 and https://github.com/mysociety/alaveteli/pull/7201/commits/a5f8bcc5aee7db18778b1d096f03db863d24540a

File: app/controllers/admin/tags_controller.rb
Comment (gbp) [reply app/controllers/admin/tags_controller.rb]: As we've discussed off-site I think the fixups address the concerns here.

File: app/views/admin/tags/show.html.erb
Comment (garethrees) [inline app/views/admin/tags/show.html.erb:18]: let's drop this `<hr>` since it feels like it separates the filters from the list the filters operate on. The two are connected (and don't have `<hr>` elsewhere)

File: app/views/admin/tags/show.html.erb
Comment (garethrees) [inline app/views/admin/tags/show.html.erb:8]: Think these buttons would be better placed above the filter form, as they're more closely related to the "tag" rather than the list of "taggables" (but appreciate it makes sense that they're under the filter tabs given where they take you).

File: app/views/admin/tags/show.html.erb
Comment (garethrees) [inline app/views/admin/tags/show.html.erb:1]: Super minor, but let's make it "Tag –" as we're talking about a singular record.

File: app/views/admin_public_body_categories/edit.html.erb
Comment (garethrees) [inline app/views/admin_public_body_categories/edit.html.erb:11]: This would make it a bit clear that the button relates to the _tag_ admin page:

```diff
diff --git a/app/views/admin_public_body_categories/_form.html.erb b/app/views/admin_public_body_categories/_form.html.erb
index 9c3b9bdcd..37234bde9 100644
--- a/app/views/admin_public_body_categories/_form.html.erb
+++ b/app/views/admin_public_body_categories/_form.html.erb
@@ -62,6 +62,10 @@
          cannot be modified.
        </span>
      <% end %>
+
+     <% if @public_body_category.category_tag %>
+       <%= link_to 'Tag admin page', admin_tag_path(tag: @public_body_category.category_tag), class: 'btn' %>
+     <% end %>
   </div>
 </div>
 
diff --git a/app/views/admin_public_body_categories/edit.html.erb b/app/views/admin_public_body_categories/edit.html.erb
index 4b5b69fdc..2cb268ef3 100644
--- a/app/views/admin_public_body_categories/edit.html.erb
+++ b/app/views/admin_public_body_categories/edit.html.erb
@@ -8,7 +8,6 @@
 
         <div class="form-actions">
           <%= f.submit 'Save', accesskey: 's', class: 'btn btn-success' %>
-          <%= link_to 'Admin page', admin_tag_path(tag: @public_body_category.category_tag), class: 'btn' %>
           <%= link_to 'Public page', list_public_bodies_by_tag_path(@public_body_category.category_tag), class: 'btn' %>
           <%= link_to 'List all', admin_categories_path, class: 'btn' %>
         </div>
```

I did have a quick faff around with [attached buttons](https://getbootstrap.com/2.2.0/base-css.html#forms:~:text=Buttons%20instead%20of%20text) but it added a more complexity because of the existing conditional. Don't think it's high value enough to worry about, but up to you.

File: app/models/concerns/taggable.rb
Comment (garethrees) [inline app/models/concerns/taggable.rb:12]: NICE! 😎

**Review by garethrees — APPROVED** [review]
Nice! Looking great! 🏷️ 

Couple of minor tweaks suggested, but no need to re-review.

File: app/models/info_request.rb
Comment (gbp) [reply app/models/info_request.rb:59]: TIL

Comment (gbp) [reply app/models/info_request.rb:59]: Ah this effects the other parts of the application such as expected params. Will revert the fixup and continue with `admin_title`.

**Pull request merged**
